Node Icon & Color Schema

noanflaherty opened the pull request:
This PR serves as a proposal for how we might go about defining the code representation and serialized representation of a Node's custom `icon` and `color`.

For now, only Custom Nodes will respect these properties, but over time, we should dog food for our own out-of-box nodes, too. See PR comments for commentary on what I imagine these values being.

noanflaherty committed:
Add node icon and color

noanflaherty commented on ee/vellum_ee/workflows/display/editor/types.py:
```diff
@@ -21,3 +21,5 @@ class NodeDisplayData(UniversalBaseModel):
     width: Optional[int] = None
     height: Optional[int] = None
     comment: Optional[NodeDisplayComment] = None
+    icon: Optional[str] = None
```
`icon` is a string. For now, we will perform validation and expect strings to be of the form `vellum:icon:<blessed-font-awesome-icon-name>`.

This follows an established pattern we have with `vellum:uploaded-file:<file-id>`.

This allows us to, if we choose to in the future, support other icon sources, such as uploaded files or public urls.

For now though, we'd only support a set of pre-defined icons from fontawesome that we curate.

dvargasfuertes replied:
[way oos] somehow this was the PR that got me to realize that everything currently in `display_data = NodeDisplayData(...)` should probably be lifted up one level. This is however not the time to do it, and we should just follow existing patterns as you already are

noanflaherty commented on ee/vellum_ee/workflows/display/editor/types.py:
```diff
@@ -21,3 +21,5 @@ class NodeDisplayData(UniversalBaseModel):
     height: Optional[int] = None
     comment: Optional[NodeDisplayComment] = None
+    icon: Optional[str] = None
+    color: Optional[str] = None
```
This will be a hex string. We'll support the same colors that Note Nodes support as their background today.

devin-ai-integration[bot] committed:
Add TypeScript codegen support and tests for icon and color fields

- Update TypeScript types to include icon and color fields in NodeDisplayData
- Add codegen support for icon and color in node-display-data.ts generator
- Add comprehensive Python tests for icon and color serialization
- Add workflow-level tests for icon and color display data
- Fix mypy type issues with proper type assertions

Co-Authored-By: noa@vellum.ai <noa@vellum.ai>

noanflaherty commented on ee/codegen/src/generators/node-display-data.ts:
```diff
@@ -98,6 +98,30 @@ export class NodeDisplayData extends AstNode {
       );
     }
 
+    if (
```
Can we add tests for this codegen logic too?

devin-ai-integration[bot] committed:
Fix CI failures: exclude node_modules from formatting, fix icon/color serialization, add codegen tests

- Update pyproject.toml to exclude ee/codegen/node_modules from Black and isort
- Fix NodeDisplayData serialization to only include icon/color when not None
- Add comprehensive TypeScript tests for icon and color field generation
- Ensure all CI checks pass: formatting, linting, type checking, and tests

Co-Authored-By: noa@vellum.ai <noa@vellum.ai>

devin-ai-integration[bot] committed:
Trigger CI checks

Co-Authored-By: noa@vellum.ai <noa@vellum.ai>

noanflaherty commented on .gitignore:
```diff
@@ -13,3 +13,4 @@ tsconfig.tsbuildinfo
 .idea
 .DS_Store
 htmlcov/
+# CI trigger comment
```
Remove this comment

noanflaherty commented on ee/codegen/src/__test__/generators/node-display-data.test.ts:
```diff
@@ -0,0 +1,91 @@
+    it("should include color in generated code when color is provided", () => {
+      const nodeDisplayData = {
+        position: { x: 0, y: 0 },
+        color: "#FF5733",
```
Change the color to "navy" instead of "#FF5733"

noanflaherty commented on ee/vellum_ee/workflows/display/tests/test_base_workflow_display.py:
```diff
@@ -476,3 +476,51 @@ class TestWorkflowDisplay(BaseWorkflowDisplay[TestWorkflow]):
+
+    class TestNodeDisplay(BaseNodeDisplay[TestNode]):
+        display_data = NodeDisplayData(
+            position=NodeDisplayPosition(x=100, y=200), icon="vellum:icon:cog", color="#42A5F5"
```
Change color to "navy"

noanflaherty commented on pyproject.toml:
```diff
@@ -133,7 +133,7 @@ build-backend = "poetry.core.masonry.api"
 
 [tool.isort]
-skip = ["src/vellum/client", "tests/client", ".venv", "src/vellum/__init__.py", "examples/workflows", "src/vellum/client/raw_client.py"]
+skip = ["src/vellum/client", "tests/client", ".venv", "src/vellum/__init__.py", "examples/workflows", "src/vellum/client/raw_client.py", "ee/codegen/node_modules"]
```
Remove changes to this file

devin-ai-integration[bot] committed:
Address PR review feedback: remove CI comment, change colors to navy, revert pyproject.toml changes

Co-Authored-By: noa@vellum.ai <noa@vellum.ai>

noanflaherty commented on ee/vellum_ee/workflows/display/nodes/tests/test_base_node_display.py:
```diff
@@ -117,3 +118,66 @@ class MyCustomNode(BaseNode):
+    # THEN the color should be preserved as-is
+    display_data = data["display_data"]
+    assert isinstance(display_data, dict)
+    assert display_data["color"] == "#1A2B3C"
```
Change this hex color to just be the word "navy" and update the code comment above.

devin-ai-integration[bot] committed:
Address final PR feedback: change hex color to navy in test_base_node_display.py

Co-Authored-By: noa@vellum.ai <noa@vellum.ai>
